tests/test_ir.py

```diff
@@ -149,6 +167,95 @@ def test_op_clone_with_regions():
     assert if2.false_region.op is not if_.false_region.op
 
 
+@irdl_op_definition
+class SuccessorOp(IRDLOperation):
+    """
+    Utility operation that requires a successor and has the IsTerminator
+    trait.
+    """
+
+    name = "test.successor_op"
+
+    successor: Successor = successor_def()
+
+    traits = frozenset([IsTerminator()])
+
+
+def test_block_branching_to_another_region_wrong():
+    """
+    Tests that an operation cannot have successors that branch to blocks of
+    another region.
+    """
+    block1 = Block([TestOp.create(), TestOp.create()])
+    region1 = Region([block1])
+
+    op0 = TestOp.create(successors=[block1])
+    block0 = Block([op0])
+    region0 = Region([block0])
+    region0 = TestOp.create(regions=[region0, region1])
+
+    outer_block = Block([region0])
+
+    with pytest.raises(
+        VerifyException,
+        match="Branching to a block of a different region",
+    ):
+        outer_block.verify()
+
+
+def test_block_not_branching_to_another_region():
+    """
+    Tests that an operation can have successors that branch to blocks of the
+    same region.
+    """
+    block0 = Block()
+
+    op0 = SuccessorOp.create(successors=[block0])
+    block1 = Block([op0])
+
+    region0 = Region([block0, block1])
+
+    region0.verify()
+
+
+def test_empty_block_with_no_parent_region_requires_no_terminator():
+    """
+    Tests that an empty block belonging no parent region requires no terminator
+    operation.
+    """
+    block0 = Block([])
+
+    block0.verify()
+
+
+def test_empty_block_with_orphan_single_block_parent_region_requires_no_terminator():
+    """
+    Tests that an empty block belonging to a single-block region with no parent
+    operation requires no terminator operation.
+    """
+    block0 = Block([])
+    region0 = Region([block0])
+
+    region0.verify()
+
+
+def test_empty_block_with_single_block_parent_region_requires_terminator():
+    """
+    Tests that an empty block belonging to a single-block region in a parent
+    operation requires terminator operation.
+
+    This test should fail once the NoTerminator functionality is implemented.
+    See https://github.com/xdslproject/xdsl/issues/1093
+    """
+    block0 = Block([])
+    region0 = Region([block0])
+    op0 = TestOp.create(regions=[region0])
+
+    # TODO single-block regions dealt when the NoTerminator trait is
```

Comment:
I don't quite like inline TODOs. I feel we have missed so many of those that we decided to start opening issues most of the time.

In this case, I would remove this TODO and the last paragraph of the description of the testcase. Then you can just fix the test in the other PR. This is still a case showing a different kind of IR structure, so I am fine with having it in, even though it is actually illegal.

Reply:
It it illegal, but not yet

I'm a bit confused regarding the TODO because there is a gh issue tracking this, and it is part of the comment (as @superlopuh had asked for), so not sure what to do here.

Reply:
This one: https://github.com/xdslproject/xdsl/issues/1093

Why mention it here again then?

Reply:
No problem!